rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -2013,17 +2037,278 @@ extern void ppu_initialize()
 		prx_list.emplace_back(&prx);
 	});
 
+	bool compile_fw = is_gui;
+
 	// Initialize preloaded libraries
 	for (auto ptr : prx_list)
 	{
-		ppu_initialize(*ptr);
+		compile_fw |= ppu_initialize(*ptr);
 	}
 
-	// Initialize SPU cache
-	spu_cache::initialize();
+	const auto thread = g_fxo->init<named_thread>("SPRX Loader"sv, [compiled_main, compile_fw, is_gui, prx_list = std::move(prx_list)]
+	{
+		// Map fixed address executables area, fake overlay support
+		const bool had_ovl = !vm::map(0x3000'0000, 0x1000'0000, 0x200).operator bool();
+		const bool ppc_seg = std::exchange(g_ps3_process_info.ppc_seg, 0x3);
+
+		const std::string firmware_sprx_path = g_cfg.vfs.get_dev_flash() + "sys/external/";
+
+		std::vector<std::string> dir_queue;
+		
+		if (compile_fw)
+		{
+			dir_queue.emplace_back(firmware_sprx_path);
+		}
+
+		if (compiled_main)
+		{
+			dir_queue.emplace_back(vfs::get(Emu.GetDir()) + '/');
+
+			if (Emu.GetCat() == "DG" && !Emu.GetTitleID().empty())
+			{
+				dir_queue.emplace_back(vfs::get("/dev_hdd0/game/") + Emu.GetTitleID() + '/');
+			}
+		}
+
+		enum class exec_type
+		{
+			none,
+			prx,
+			ovl,
+			//elf,
+		};
+	
+		std::vector<std::pair<std::string, exec_type>> file_queue;
+		file_queue.reserve(2000);
+
+		// Find all .sprx files recursively (TODO: process .mself files)
+		for (usz i = 0;; i++)
+		{
+			if (i >= dir_queue.size())
+			{
+				g_progr_ftotal = file_queue.size();
+				break;
+			}
+
+			if (Emu.IsStopped())
+			{
+				file_queue.clear();
+				break;
+			}
+
+			sys_log.notice("Scanning directory: %s", dir_queue[i]);
+
+			for (auto&& entry : fs::dir(dir_queue[i]))
+			{
+				if (Emu.IsStopped())
+				{
+					i = -1;
+					file_queue.clear();
+					break;
+				}
+
+				if (entry.is_directory)
+				{
+					if (entry.name != "." && entry.name != "..")
+					{
+						dir_queue.emplace_back(dir_queue[i] + entry.name + '/');
+					}
+
+					continue;
+				}
+
+				// Check .sprx, .prx, .self, .elf filenames
+				if (exec_type result = [&]()
+				{
+					std::string filename = fmt::to_upper(entry.name);
+
+					if (filename.ends_with(".SPRX"sv) || filename.ends_with(".PRX"sv))
+					{
+						return exec_type::prx;
+					}
+
+					if (filename.ends_with(".SELF"sv) || filename.ends_with(".ELF"sv))
```

Comment:
Or it could be another "main executable"

Reply:
I implemented basic error checking for it in ppu_load_overlay.